arangod/Cluster/ServerState.cpp

```diff
@@ -661,10 +693,32 @@ bool ServerState::registerAtAgencyPhase2(AgencyComm& comm) {
       FATAL_ERROR_EXIT();
     }
 
-    auto result = comm.setValue(currentServersRegisteredPref + _id, builder.slice(), 0.0);
+    // If we generated a new UUID, this *must not* exist in the Agency, so we
+    // should fail to register.
+    std::vector<AgencyPrecondition> pre;
+    if (!hadPersistedId) {
+      pre.emplace_back(AgencyPrecondition(rebootIdPath, AgencyPrecondition::Type::EMPTY, true));
+    }
+
+    AgencyWriteTransaction trx(
+        {AgencyOperation(serverRegistrationPath, AgencyValueOperationType::SET,
+                         builder.slice()),
+         AgencyOperation(rebootIdPath, AgencySimpleOperationType::INCREMENT_OP),
+         AgencyOperation("Current/Version", AgencySimpleOperationType::INCREMENT_OP)},
+        pre);
+
+    auto result = comm.sendTransactionWithFailover(trx, 0.0);
 
     if (result.successful()) {
-      return true;
+      uint64_t rebootId;
+
+      // If reading back the rebootId is not successful, we might
+      // increment it again above, so maybe we need to make a separate
+      // retry loop here?
+      if (readRebootIdFromAgency(comm, rebootId) == 0) {
```

Comment:
We might want a retry loop either in `readRebootIdFromAgency` or here around it, maybe even with an additional increment. Imagine a situation where the agency is temporarily unavailable and a server restarts. Then the server would eventually start but think that its rebootId is 0. In any case, if we update `rebootIds` for all servers in `ClusterInfo` and do not use the `rebootId` here from `ServerState`, we can even remove the reading here, potentially. What do others think?

Reply:
I am tempted to leave it here just to make sure it was set.

Reply:
I added and removed the code for reading `rebootId` back into `ServerState` at least twice: I eventually put it back in, because I thought it'd be useful to make sure it was set (as you say).